src/util/uploadExternalFormatFile.tsx

```diff
@@ -0,0 +1,150 @@
+  imageFile: File | null;
+  name: string;
+  pool: string;
+  formatConversion: boolean;
+  virtioConversion: boolean;
+  architecture: string;
+}
+
+export const FILE_CHUNK_SIZE = 1024 * 1024;
+
+export const supportedVMArchOptions = (hostArchitectures: string[]) => {
+  const vmArchitectureOptions = [
+    { value: "x86_64", label: "x86_64" },
+    { value: "aarch64", label: "aarch64" },
+    { value: "ppc64le", label: "ppc64le" },
+    { value: "s390x", label: "s390x" },
+  ];
+
+  return vmArchitectureOptions.filter((arch) =>
+    hostArchitectures.includes(arch.value),
+  );
+};
+
+export const uploadExternalFormatFilePayload = (
+  values: UploadExternalFormatFileFormValues,
+) => {
+  const fileSize = values.imageFile?.size || 0;
+  const conversionOptions = [];
+
+  if (values.formatConversion) {
+    conversionOptions.push("format");
+  }
+
+  if (values.virtioConversion) {
+    conversionOptions.push("virtio");
+  }
+
+  const instanceArgs = {
+    source: {
+      type: "conversion",
+      mode: "push",
+      conversion_options: conversionOptions,
+    },
+    devices: {
+      root: {
+        path: "/",
+        type: "disk",
+        pool: values.pool,
+      },
+    },
+    type: "virtual-machine",
+    name: values.name,
+    sourceDiskSize: fileSize,
+    architecture: values.architecture,
+  };
+
+  return JSON.stringify(instanceArgs);
+};
+
+export const sendFileByWebSocket = (
+  ws: WebSocket | null,
+  file: File | null,
+  chunkSize: number,
+  onProgress?: (current: number, total: number) => void,
+  onError?: (error: Error) => void,
+) => {
+  if (ws && ws.readyState === WebSocket.OPEN) {
+    const reader = new FileReader();
+    let sentFileSize = 0;
+
+    const readNextFileChunk = () => {
+      if (!file) {
+        return;
+      }
+      const chunk = file.slice(sentFileSize, sentFileSize + chunkSize);
+      if (!chunk.size) {
+        // NOTE: if we close the connection without a code, the default 1005 is used which will cause backend to throw an error
+        ws.close(1000, "File upload complete");
+        return;
+      }
+
+      reader.readAsArrayBuffer(chunk);
+    };
+
+    reader.onload = (event) => {
+      if (!file) {
+        return;
+      }
+
+      const result = event.target?.result;
+      let data: ArrayBuffer | ArrayBufferLike;
+      if (result && typeof result === "string") {
+        // Convert string to ArrayBuffer
+        data = new TextEncoder().encode(result).buffer;
+      } else {
+        data = result as ArrayBuffer;
+      }
+
+      ws.send(data);
+
+      sentFileSize += data.byteLength;
+      if (onProgress) {
+        onProgress(sentFileSize, file.size);
+      }
+
+      readNextFileChunk();
+    };
+
+    reader.onerror = (e) => {
+      if (onError) {
+        onError(
+          e.target?.error
+            ? e.target.error
+            : new Error("Failed to read read file"),
+        );
+      }
+    };
+
+    readNextFileChunk();
+  }
+};
```

Comment:
I found this hard to understand. Maybe splitting the responsibilities into a file reader, that just reports chunks and upload state, and another part that is hooking that up with the socket? Or maybe that is even worse, and the complexity is inherent.

Reply:
I agree this is not the easiest code to understand but I am also not sure if we can split it like that since the socket and reader logic are tightly coupled. perhaps I can add some comments above the function describing how it works?

Reply:
Yeah, a comment might help.

Reply:
I've added some comments explaining the function.